File: javascript/array-hasing/two-sum.js
// Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

// You may assume that each input would have exactly one solution, and you may not use the same element twice

//output: [index, index]

function twoSum(nums, target) {
    const map = new Map()

    // map every number until we find the diff
    // initial mapping 
    map.set(nums[0], 0)

    //go through nums array
    for (let i = 1; i < nums.length; i++) {
        //compute the number we search for
        const diff = target - nums[i]

        //search for diff
        if (map.has(diff)) {
            return [map.get(diff), i]
        }

        // map number with his index
        map.set(nums[i], i)
    }
}

// testing
console.log(twoSum([2,7,11,15], 9)) // expected [0,1]

File: python/array-hashing/two_sum.py
# Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

# You may assume that each input would have exactly one solution, and you may not use the same element twice.

# You can return the answer in any order.

#output: [index, index]

def twoSum(nums: list, target: int) -> list:
    map = {}

    for i, n in enumerate(nums):
        # compute the number we search for
        diff = target - n

        # search for diff
        if diff in map:
            return [map[diff], i]

        # map number with his index
        map[n] = i


# testing
print(twoSum([3,2,4], 6)) # expected [1,2]